File: src/main/java/PythonSol/LinearDataStructures/Lists/AssociationList/AssociationList.py
from typing import Any

class Node:
    def __init__(self, key: int, value: Any, next: 'Node' = None) -> None:
        self.key:int = value
        self.value: Any = value
        self.next: 'Node' = next


class AssociationList:
    def __init__(self):
        self.head = None
        self.size = 0

    def __len__(self) -> int:
        return self.size
    
    def append(self, key: int, value: Any) -> None:
        if self.head == None:
            self.head = Node(key, value)
        else:
            curr: Node = self.head
            while curr.next != None:
                curr = curr.next

            new_node: Node = Node(key, value)
            curr.next = new_node

        self.size += 1

    def delete(self) -> None:
        if self.head == None:
            return

        if self.head.next == None:
            self.head = None
            return

        curr: Node = self.head
        while curr.next.next != None:
            curr = curr.next

        curr.next = None

        self.size -= 1

    def delete_index(self, index: int) -> None:
        if index < 0 or index >= self.size:
            raise Exception("Index out of bounds")

        if self.head == None:
            return

        if index == 0:
            next_node: Node = self.head.next
            del self.head
            self.head = next_node
            self.size -= 1
            return

        counter = 0
        prev: Node = None
        curr: Node = self.head

        while counter != index:
            prev = curr
            curr = curr.next
            counter += 1

        prev.next = curr.next
        del curr
        self.size -= 1

    def search(self, key: int) -> Any:
        if self.head == None:
            raise Exception("The linked list is empty.")

        curr: Node = self.head
        while curr != None:
            if curr.key == key:
                return curr.value

            curr = curr.next

        raise Exception("Value not found.")

    def print_list(self) -> None:
        if self.head == None:
            print("Empty list.")
            return

        curr: Node = self.head
        while curr != None:
            print(curr.value, end = " ")
            curr = curr.next

        print()

File: src/main/java/CCPPSol/LinearDataStructures/Lists/AssociationList/AssociationList.cpp
#include <iostream>

using namespace std;

class Node {
public:
    int key;
    int value;
    Node* next;

    Node(int key, int value) {
        this->key = key;
        this->value = value;
    }
};

class AssociationList {
private:
    int size = 0;
public:
    Node* head;

    int get_size() {
        return this->size;
    }

    void append(int kye, int value) {
        if(this->head == NULL) {
            this->head = new Node(key, value);
        } else {
            Node* current = this->head;

            while(current->next != NULL) {
                current = current->next;
            }
            
            current->next = new Node(key, value);
        }

        this->size += 1;
    }

    void delete_last() {
        if (this->head == NULL) {
            cout << "The linked list is empty." << endl;
            return;
        }

        if (this->head->next == NULL) {
            this->head = NULL;
            return;
        }

        Node* current = this->head;
        while (current->next->next != NULL) {
            current = current->next;
        }

        current->next = NULL;
        this->size -= 1;
    }

    void delete_index(int index) {
        if (index < 0 || index >= this->size) {
            return;
        }

        if(index == 0) {
            Node* current_head = this->head;
            Node* next = this->head->next;

            this->head = next;
            this->size -= 1;
            return;
        }

        int counter = 0;
        Node* previous = NULL;
        Node* current = this->head;

        while (counter != index) {
            previous = current;
            current = current->next;
            counter += 1;
        }

        previous->next = current->next;
        this->size -= 1;
    }

    int search(int key) {
        if (this->head == NULL) {
            cout << "The linked list is empty." << endl;
            return NULL;
        }

        Node* current = this->head;

        while (current != NULL) {
            if(current->key == key) {
                return current->value;
            }

            current = current->next;
        }


        cout << "No node found that contains the given value inside of the linked list." << endl;
        return NULL;
    }

    void print_list() {
        if(this->head == NULL) {
            cout << "The linked list is empty." << endl;
            return;
        }

        Node* current = this->head;
        while (current != NULL) {
            cout << current->value << " ";
            current = current->next;
        }

        cout << " " << endl;
    }
};